Giulio/code review

caesar2164 opened the pull request:
@lydiali, @z4y4ts - PR so we can see code changes

FYI @stvstnfrd

z4y4ts committed:
First draft

COTOHA committed:
Update README.md

z4y4ts committed:
Add author_view.html to display xblock status to staff

z4y4ts committed:
Add student_view.html to check if user registered in edx adapt

z4y4ts committed:
Add skills field to xblock

z4y4ts committed:
Improve logging and strip trailing / fron api url when pass it to student_view

z4y4ts committed:
Improve code style

z4y4ts committed:
Delete left-over line

z4y4ts committed:
Merge pull request #1 from raccoongang/dev

Basic functionality

z4y4ts committed:
Fix user configuration call to edx-adapt

z4y4ts committed:
Make anonymous_student_id and course_id compatible with capa problems JS

* Use course non-specific anonymous_student_id
* Use short version of course_id without organization & course run

Commit:
Add documentation in README file

* XBlock fields description
* User registration flow description
* Brief Edx Adapt logic overview

z4y4ts committed:
Add EdxAdaptXBlock.ugettext() for backward compatibility and fix typo

z4y4ts committed:
Improve XBlock params edit page

Add helpful display names and help text.
Hide unneeded field

idegtiarov committed:
Add assistance messageis in case adapt enrollment successes or fails

caesar2164 commented on setup.py:
```diff
@@ -0,0 +1,42 @@
+    name='edxadapt-xblock',
+    version='0.1',
+    description='edxadapt XBlock',   # TODO: write a better description.
+    license='UNKNOWN',          # TODO: choose a license: 'AGPL v3' and 'Apache 2.0' are popular.
```
this should DEFINITELY be AGPL v3 BEFORE we can install it

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+    var params = {{params}};
+    var skills = {{skills}};
+
+    var registerUser = function(){
```
lets define these using function declarations instead of function expressions <http://stackoverflow.com/questions/336859/javascript-function-declaration-syntax-var-fn-function-vs-function-fn> to take advantage of hoisting...

i.e.:
```javascript
function registerUser() {
    // Create user
    $.ajax({
        url: apiBaseUrl + '/course/' + courseId + '/user',
        type: 'POST',
        contentType: 'application/json',
        data: JSON.stringify({'user_id': studentId}),
    })
    .done(function() {
        console.log("User Created");
        configureUser();
    })
    .fail(function() {
        console.log("Failed to create user");
        setStatusNok();
    });
};
```

z4y4ts replied:
Good point.

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+        .fail(function() {
+            console.log("Failed to create user");
+            setStatusNok();
+        })
```
missing semicolon

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+                console.log('Failed to configure skill "'+skill+'" for user "'+studentId+'"');
+                setStatusNok();
+            });
+        })
```
missing semicolon

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+        })
+    };
+
+    var configureUser = function() {
```
see comment above about function declarations.

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+        })
+    };
+
+    var registerUserInEdxAdapt = function() {
```
see comment above about function declarations.

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+            console.log("Failed to register student in EdxAdapt", jqXHR.responseText);
+        });
+    }
+    var setStatusOk = function() {
```
see comment above about function declarations.

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+    var setStatusOk = function() {
+        $('#status_ok').css('visibility', 'visible');
+    }
+    var setStatusNok = function() {
```
see comment above about function declarations.

caesar2164 commented on edxadapt/static/js/src/edxadapt.js:
```diff
@@ -0,0 +1,82 @@
+    var apiBaseUrl = '{{edx_adapt_api_url}}';
+    var courseId = '{{course_id}}';
+    var params = {{params}};
+    var skills = {{skills}};
```
I know why you have to do this, but add comment explaining the double curly brackets

caesar2164 replied:
Actually could these variable values be passed through and read from the HTML as "data-" attributes so that this JS file can be minified and cached?

caesar2164 commented on edxadapt/static/html/student_view.html:
```diff
@@ -0,0 +1,10 @@
+</div>
+<div id="status_nok">
+    <i class="fa fa-exclamation-circle" aria-hidden="true" style="color:red;"></i> There was a technical issue while we were configuring your adaptive learning session. Please try to refresh this page or contact technical staff if problem persists.
+</div>
```
since we have access to CSS, let's put the color into a CSS class:

HTML
```HTML
<div id="status_ok">
    <i class="fa fa-check-circle green" aria-hidden="true"></i>
</div>
<div id="status_nok">
    <i class="fa fa-exclamation-circle red" aria-hidden="true"></i> There was a technical issue while we were configuring your adaptive learning session. Please try to refresh this page or contact technical staff if problem persists.
</div>
```

CSS
```CSS
.green {
    color: green;
}
.red {
    color: red;
}
```

z4y4ts replied:
Makes sense.

caesar2164 commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,137 @@
+    params = Dict(
+        default={'pg': 0.25, 'ps': 0.25, 'pi': 0.1, 'pt': 0.5, 'threshold': 0.99},
+        display_name=_('BKT params'),
+        help=_('Barameters for Bayesian Knowledge Tracing (BKT) model'),
```
"Parameters"?

caesar2164 commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,137 @@
+    has_author_view = True
+
+    display_name = String(
+        default='Adaptive Learning',
```
maybe 'Adaptive Learning Excercise'?

that way if someone doesn't change the display name it makes more sense to learners?

z4y4ts replied:
This field actually is not being displayed to learners. The only thing they see in this xblock is a green or red status mark.

It does make sense to change xblock title to something more descriptive though.
Like `"Adaptive Learning Enrollment"` since this is what the xblock actually does.

caesar2164 replied:
That works for me!

caesar2164 commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,137 @@
+
+        return anonymous_id_for_user(
+            User.objects.get(
+                id=self.runtime.user_id), None
```
I'm not ecstatic about having to use the non-anon user id to get the anon one. I think xblocks might have a built in function to get the anon-id?

caesar2164 replied:
here's what I was thinking: https://groups.google.com/forum/#!topic/edx-code/ryPw7a-tK_g

```python
@property
    def student_id(self):
        return self.xmodule_runtime.anonymous_student_id
```

z4y4ts replied:
I'm not happy about this part too. And yes, runtime does provide `anonymous_student_id`.

Unfortunately it's a "course specific" anonymous id. Whereas we need "course non specific" anonymous id, because only this type of anonymous id is available in a capa-problems context, when students solve adaptive problems.

Here's comparison:
```python
# Provided by XBlock runtime
ipdb> self.runtime.anonymous_student_id
'58cfd06b8e049a0e599c9eea166f8857'

# Course non-specific anonymous id
ipdb> anonymous_id_for_user(User.objects.get(id=self.runtime.user_id), None)
'66c6d171816ef81dc88859f0e87ae732'

# Course specific anonymous id
ipdb> anonymous_id_for_user(User.objects.get(id=self.runtime.user_id), self.course_id)
'58cfd06b8e049a0e599c9eea166f8857'
```

caesar2164 replied:
damn, ok sadly this probably has to stay as it is

stvstnfrd replied:
@z4y4ts Can you clarify why you need the "course non specific" anonymous id?
It's generally considered an anti-pattern for any XBlock to directly access Django models.

z4y4ts replied:
@stvstnfrd Our problems are Custom JavaScript problems. They are provided by student anonymous id within their context, which is "course non specific" anonymous id. So we can communicate only "course non specific" anon. id when student solve problems. Hence we use it during student registration phase.

But we've updated the code to not use `student.models.User` directly.

stvstnfrd replied:
@z4y4ts Thanks for the explanation and cleanup.

caesar2164 commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,137 @@
+                    'static/js/src/edxadapt.js'
+                )).render(Context({
+                    'anonymous_student_id': anonymous_student_id,
+                    'course_id': self.get_course_id(),
```
this seems like unnecessary redirection. Couldn't we just define it like this:

```python
'course_id': self.course_id.course,
```

z4y4ts replied:
The idea was to put an explanation why it's `self.course_id.course` in a function docstring. But yes, this can be simplified.

caesar2164 replied:
Let's keep this as-is, the named function and the docstring really helps to understand what is going on.

caesar2164 reviewed (commented):
This is looking very good, just have some comments.

caesar2164 commented:
@z4y4ts when making changes, just add them to your master, this branch just tracks that. (I'll make sure to update this with any changes you make)

z4y4ts committed:
Create LICENSE

idegtiarov committed:
Fix typos and improve functions' declarations in js

idegtiarov committed:
Add documentation

Add description for new editable field 'Edx Adapt Course ID' and figures with all
xblock's editable fields.

idegtiarov committed:
Improve img and text in the doc

Commit:
Add editable course_id field

This field is required when course has more than one sections with adaptive problems to distinguish
adaptive sub-course on edx-adapt side.

Commit:
Change from one-by-one skill registration to bulk processing (#6)

Change from one-by-one skill registration to bulk processing

Commit:
Fix doc strings in js file

Commit:
Emit event after successful student enrollment in EdxAdapt

Event passes url to the first first problem to solve and is used to enable "next" buttons.

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        'edx_adapt_api_url',
+        'success_registration_msg',
+        'fail_registration_msg',
+        'edx_adapt_course_id'
```
Add trailing comma

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+import logging
+import pkg_resources
+
+from django.template import Template, Context
```
Remove unused import

idegtiarov replied:
done

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+    has_author_view = True
+
+    display_name = String(
+        default='Adaptive Learning Enrollment',
```
Wrap string for translation

kluo replied:
Also could you add display name and help parameters

idegtiarov replied:
done

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+            'd to h', 'histogram', 'None'
+        ],
+        display_name=_('Course skills list'),
+        scope=Scope.content,
```
Can you confirm the Scope.content gives the intended effect for these attributes? From http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/concepts/fields.html#predefined-scopes it would appear that this would mean anytime these attributes are changed they are reflected across all instance of the XBlock, regardless of the course.

idegtiarov replied:
Good catch. I improve scope from content to settings

z4y4ts replied:
Actually, `Scope.settings` does fits better. Because EdxAdapt treats every course run and usage separately.

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        default={'pg': 0.25, 'ps': 0.25, 'pi': 0.1, 'pt': 0.5, 'threshold': 0.99},
+        display_name=_('BKT params'),
+        help=_('Parameters for Bayesian Knowledge Tracing (BKT) model'),
+        scope=Scope.content
```
Add trailing comma

idegtiarov replied:
done

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        help=_(
+            'List of skills of this course. Each problem addresses certain skill. '
+            'Special skill "None" is used for those problems which belong to none.'
+        )
```
Trailing comma

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        default='',
+        display_name=_('Edx Adapt API base URL'),
+        scope=Scope.content,
+        help=_('Edx Adapt API base URL, e.g. https://edx-adapt.example.com:443/api/v1')
```
Trailing comma

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+            "Stanford+PS01+2017. In case when adaptive problems are spread between several course's section, Course ID "
+            "should be set manually with adding section name where problems are placed, e.g. "
+            "University+Course_name+course_run:section_name. E.g. Stanford+PS01+2017:introduction"
+        )
```
Trailing comma

idegtiarov replied:
done

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+
+    student_is_registered = Boolean(
+        default=False,
+        scope=Scope.preferences
```
Trailing comma

idegtiarov replied:
Done

kluo commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        'edx_adapt_api_url',
+        'success_registration_msg',
+        'fail_registration_msg',
+        'edx_adapt_course_id'
```
Trailing comma

idegtiarov replied:
done

stvstnfrd commented on edxadapt/edxadapt.py:
```diff
@@ -0,0 +1,172 @@
+        help=_('Edx Adapt API base URL, e.g. https://edx-adapt.example.com:443/api/v1')
+    )
+
+    edx_adapt_course_id = String(
```
This is named "_course_id", but the help text documentation implies that additional information (i.e. `section_name`) could also be included.
Could you clarify the use of this variable and its intent?

idegtiarov replied:
course_id is used as ID for the course instance on EdxAdapt side. If we have more than one adaptive section in the course each of them setup in EdxAdapt as a separate instance and to distinguish them we add section_name to the course_id.

stvstnfrd replied:
Thanks. I understand that/why you need it, I'm just not keen on the naming convention, since `course_id` means something very specific w/in edX.
That said, I can live with it for now.

Commit:
Fix fields scopes, update edx_adapt_course_id help text

Also some code cleanup.

Commit:
Decouple XBlock from student.models.User

By using XBlock.system.get_real_user() to get user object by available anonymous id.
Also prevent student.models.anonymous_id_for_user() saving AnonymousUserId.
